<a href="https://colab.research.google.com/github/HadeerMahdy19/pandas_playground/blob/main/Clean_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

### 1- Read file from my workspace 

In [9]:
data_path = '/content/sample_data/countries of the world.csv'  
df = pd.read_csv(data_path ,decimal=',')

print('The shape of the dataset is:', df.shape)

The shape of the dataset is: (229, 20)


## * The file contains 229 rows and 20 columns


### 2- Display first few rows 

In [10]:
df.head()

Country                               Region  Population  \
0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1         Albania   EASTERN EUROPE                          3581655   
2         Algeria   NORTHERN AFRICA                        32930091   
3  American Samoa   OCEANIA                                   57794   
4         Andorra   WESTERN EUROPE                            71201   

   Area (sq. mi.)  Pop. Density (per sq. mi.)  Coastline (coast/area ratio)  \
0          647500                        48.0                          0.00   
1           28748                       124.6                          1.26   
2         2381740                        13.8                          0.04   
3             199                       290.4                         58.29   
4             468                       152.1                          0.00   

   Net migration  Infant mortality (per 1000 births)  GDP ($ per capita)  \
0          23.06                              163.07               700.0   
1          -4.93                               21.52              4500.0   
2          -0.39                               31.00              6000.0   
3         -20.71                                9.27              8000.0   
4           6.60                                4.05             19000.0   

   Literacy (%)  Phones (per 1000)  Arable (%)  Crops (%)  Other (%)  Climate  \
0          36.0                3.2       12.13       0.22      87.65      1.0   
1          86.5               71.2       21.09       4.42      74.49      3.0   
2          70.0               78.1        3.22       0.25      96.53      1.0   
3          97.0              259.5       10.00      15.00      75.00      2.0   
4         100.0              497.2        2.22       0.00      97.78      3.0   

   Birthrate  Deathrate  Agriculture  Industry  Service  
0      46.60      20.34         0.38      0.24     0.38  
1      15.11       5.22       232.00    188.00   579.00  
2      17.14       4.61       101.00      0.60   298.00  
3      22.46       3.27          NaN       NaN      NaN  
4       8.71       6.25          NaN       NaN      NaN

### 3- Search for empty rows in each column 

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 229 entries, 0 to 228
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             229 non-null    object 
 1   Region                              229 non-null    object 
 2   Population                          229 non-null    int64  
 3   Area (sq. mi.)                      229 non-null    int64  
 4   Pop. Density (per sq. mi.)          229 non-null    float64
 5   Coastline (coast/area ratio)        229 non-null    float64
 6   Net migration                       226 non-null    float64
 7   Infant mortality (per 1000 births)  226 non-null    float64
 8   GDP ($ per capita)                  228 non-null    float64
 9   Literacy (%)                        211 non-null    float64
 10  Phones (per 1000)                   225 non-null    float64
 11  Arable (%)                          227 non-n

In [12]:
round(df.isnull().sum(axis=0)*100/df.shape[0],2)

Country                               0.00
Region                                0.00
Population                            0.00
Area (sq. mi.)                        0.00
Pop. Density (per sq. mi.)            0.00
Coastline (coast/area ratio)          0.00
Net migration                         1.31
Infant mortality (per 1000 births)    1.31
GDP ($ per capita)                    0.44
Literacy (%)                          7.86
Phones (per 1000)                     1.75
Arable (%)                            0.87
Crops (%)                             0.87
Other (%)                             0.87
Climate                               9.61
Birthrate                             1.31
Deathrate                             1.75
Agriculture                           6.55
Industry                              6.99
Service                               6.55
dtype: float64

### 4- Find duplicated rows for each column

In [15]:
df[df.duplicated()]

Country                               Region  Population  \
132         Australia   OCEANIA                                20264082   
228  Papua New Guinea   OCEANIA                                 5670544   

     Area (sq. mi.)  Pop. Density (per sq. mi.)  Coastline (coast/area ratio)  \
132         7686850                         2.6                          0.34   
228          462840                        12.3                          1.11   

     Net migration  Infant mortality (per 1000 births)  GDP ($ per capita)  \
132           3.98                                4.69             29000.0   
228           0.00                               51.45              2200.0   

     Literacy (%)  Phones (per 1000)  Arable (%)  Crops (%)  Other (%)  \
132         100.0              565.5        6.55       0.04      93.41   
228          64.6               10.9        0.46       1.44      98.10   

     Climate  Birthrate  Deathrate  Agriculture  Industry  Service  
132      1.0      12.14       7.51         38.0     262.0      0.7  
228      2.0      29.36       7.25        353.0     381.0    266.0

## * File contains 2 duplicated rows

### 5- Fill empty rows of numerical data with median

In [18]:
null_cols = ['Net migration', 'Infant mortality (per 1000 births)', 'GDP ($ per capita)', 'Literacy (%)', 'Phones (per 1000)', 'Arable (%)', 'Crops (%)' ,'Other (%)' ,'Climate' ,'Birthrate', 'Deathrate', 'Agriculture', 'Industry', 'Service']
for c in null_cols:
    df[c].fillna(df[c].median(), inplace=True)

### 6- Validate that empty rows are filled

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 229 entries, 0 to 228
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Country                             229 non-null    object 
 1   Region                              229 non-null    object 
 2   Population                          229 non-null    int64  
 3   Area (sq. mi.)                      229 non-null    int64  
 4   Pop. Density (per sq. mi.)          229 non-null    float64
 5   Coastline (coast/area ratio)        229 non-null    float64
 6   Net migration                       229 non-null    float64
 7   Infant mortality (per 1000 births)  229 non-null    float64
 8   GDP ($ per capita)                  229 non-null    float64
 9   Literacy (%)                        229 non-null    float64
 10  Phones (per 1000)                   229 non-null    float64
 11  Arable (%)                          229 non-n

## * Empty numerical cells are filled with median

### 7- Remove duplicates

In [21]:
df.drop_duplicates(inplace =True)

8- Validate that file contain no duplicates 

In [22]:
df[df.duplicated()]

Empty DataFrame
Columns: [Country, Region, Population, Area (sq. mi.), Pop. Density (per sq. mi.), Coastline (coast/area ratio), Net migration, Infant mortality (per 1000 births), GDP ($ per capita), Literacy (%), Phones (per 1000), Arable (%), Crops (%), Other (%), Climate, Birthrate, Deathrate, Agriculture, Industry, Service]
Index: []

## * Duplicates are removed

7- Save fixed data in a csv file

In [23]:
df.to_csv('fixed_data.csv')